x-pack/plugins/enterprise_search/server/lib/ml_inference_pipeline/get_inference_pipelines.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IngestPipeline, IngestProcessorContainer } from '@elastic/elasticsearch/lib/api/types';
+import { ElasticsearchClient } from '@kbn/core/server';
+import { MlTrainedModels } from '@kbn/ml-plugin/server';
+
+/**
+ * Gets all ML inference pipelines. Redacts trained model IDs in those pipelines which reference
+ * a model inaccessible in the current Kibana space.
+ * @param esClient the Elasticsearch Client to use to fetch the errors.
+ * @param trainedModelsProvider ML trained models provider.
+ */
+export const getMlInferencePipelines = async (
+  esClient: ElasticsearchClient,
+  trainedModelsProvider: MlTrainedModels | undefined
+): Promise<Record<string, IngestPipeline>> => {
+  if (!trainedModelsProvider) {
+    return Promise.reject(new Error('Machine Learning is not enabled'));
+  }
+
+  // Fetch all ML inference pipelines and trained models that are accessible in the current
+  // Kibana space
+  const [fetchedInferencePipelines, trainedModels] = await Promise.all([
+    esClient.ingest.getPipeline({
+      id: 'ml-inference-*',
+    }),
+    trainedModelsProvider.getTrainedModels({}),
+  ]);
+  const accessibleModelIds = Object.values(trainedModels.trained_model_configs).map(
+    (modelConfig) => modelConfig.model_id
+  );
+
+  // Process pipelines: check if the model_id is one of the redacted ones, if so, redact it in the
+  // result as well
+  const inferencePipelinesResult: Record<string, IngestPipeline> = {};
+  Object.entries(fetchedInferencePipelines).forEach(([name, inferencePipeline]) => {
+    inferencePipelinesResult[name] = {
+      ...inferencePipeline,
+      processors: inferencePipeline.processors?.map((processor) =>
+        redactModelIdIfInaccessible(processor, accessibleModelIds)
+      ),
+    };
+  });
+
+  return Promise.resolve(inferencePipelinesResult);
```

Comment:
Do we want `Promise.resolve` here? It doesn't _seem_ wrong, but I also noticed that basically none of the other functions called from the routes files do this. Is it just optional?

Reply:
Not sure how TS deals with this. Since the return type of this function is `Promise<Record<string, IngestPipeline>>` I'd expect we needed to adhere to that, but I see it works with just `return inferencePipelinesResult` too.
I'll remove `Promise.resolve()` in a follow-up PR.

Reply:
Ah, it looks like that's what `async` does. It wraps the return value in a promise 👍

Reply:
Yeah all of the Promise stuff in here is unnecessary (and very error-prone). Async functions can just be coded as synchronous functions (with `await` in front of async calls) without needing to Promise-wrap everything. 

For future reference: if you see a common pattern in the codebase and notice yourself doing something completely different, it's a good idea to get some eyes on the code from folks who are more experienced with the codebase. There's usually a reason for these conventions :)

Reply:
Addressed in #143967.
Will do - in this case I missed the pattern in the codebase, but will keep an eye out for it.